_posts/2020-05-18-websocket-fallback-sockjs.md

```diff
@@ -0,0 +1,329 @@
+
+Endpointy nie są kompatybilne ze sobą, ponieważ:
+* WebSocket-owy wystawia końcówkę (np. `/endpoint`) dla połączenia WebSocket
+* a SockJs wystawia końcówkę (HTTP, np. `/endpoint`) w celu negocjacji parametrów połączenia,<br />
+  chociaż klient również ma możliwość połączenia się za pomocą WebSocket bez instalowania SockJS wykorzystując do tego `/endpoint/websocket`
+
+## Praktyka
+Ostatecznie został wybrany SockJS z powodu prostoty wykorzystania go w istniejącym springowym projekcie.
+
+### Wersja przed poprawkami z użyciem czystego WebSocket
+#### Backend
+* pom.xml
+{% highlight xml %}
+<dependencies>
+    <dependency>
+        <groupId>org.springframework.boot</groupId>
+        <artifactId>spring-boot-starter-websocket</artifactId>
+    </dependency>
+</dependencies>
+{% endhighlight %}
+
+* WebSocketConfig.java
+{% highlight java %}
+@Configuration
+@EnableWebSocket
+public class WebSocketConfig implements WebSocketConfigurer {
+    private SocketHandler socketHandler;
+
+    public WebSocketConfig(SocketHandler socketHandler) {
+        this.socketHandler = socketHandler;
+    }
+
+    @Override
+    public void registerWebSocketHandlers(WebSocketHandlerRegistry registry) {
+        registry.addHandler(socketHandler, "/signaling");
+    }
+}
+{% endhighlight %}
+
+#### Frontend
+* web-socket-manager.ts
+{% highlight typescript %}
+export class WebSocketManager {
+    private webSocket?: WebSocket;
+    constructor() {
+        this.webSocket = this.initWebSocket();
+    }
+     
+    initWebSocket(): WebSocket {
+        return new WebSocket('https://localhost/signaling');
+    }
+}
+{% endhighlight %}
+
+
+### Wersja po poprawkach z użyciem SockJS
+* pom.xml (bez zmian)
+{% highlight diff %}
+<dependencies>
+    <dependency>
+        <groupId>org.springframework.boot</groupId>
+        <artifactId>spring-boot-starter-websocket</artifactId>
+    </dependency>
+</dependencies>
+{% endhighlight %}
+
+* WebSocketConfig.java
+{% highlight diff %}
+@Configuration
+@EnableWebSocket
+public class WebSocketConfig implements WebSocketConfigurer {
+    private SocketHandler socketHandler;
+
+    public WebSocketConfig(SocketHandler socketHandler) {
+        this.socketHandler = socketHandler;
+    }
+
+    @Override
+    public void registerWebSocketHandlers(WebSocketHandlerRegistry registry) {
+-       registry.addHandler(socketHandler, "/signaling");
++       registry.addHandler(socketHandler, "/signaling")
++           .withSockJS();
+    }
+}
+{% endhighlight %}
+
+* web-socket-manager.ts
+{% highlight diff %}
++import * as SockJS from 'sockjs-client';
++
+ export class WebSocketManager {
+    private webSocket?: WebSocket;
+    constructor() {
+        this.webSocket = this.initWebSocket();
+    }
+    
+    initWebSocket(): WebSocket {
+-      return new WebSocket('https://localhost/signaling');
++      return new SockJS('https://localhost/signaling');
```

Comment:
dajmy produkcyjne przykłady zamiast tych localhostów

teraz to wygląda jak turbo poc a przecież używamy tego w faktycznie deployowanym projekcie